govcd/admincatalog.go

```diff
@@ -129,3 +134,398 @@ func (cat *AdminCatalog) PublishToExternalOrganizations(publishExternalCatalog t
+			if err != nil {
+				return nil, err
+			}
+			return cat.AdminCatalog.Tasks, nil
+		})
+		return err
+	}
+	return nil
+}
+
+// Sync synchronises a subscribed AdminCatalog
+func (cat *AdminCatalog) Sync() error {
+	// if the catalog was not subscribed, return
+	if cat.AdminCatalog.ExternalCatalogSubscription == nil || cat.AdminCatalog.ExternalCatalogSubscription.Location == "" {
+		return nil
+	}
+	// The sync operation is only available for Catalog, not AdminCatalog.
+	// We use the embedded Catalog object for this purpose
+	catalogHref, err := cat.GetCatalogHref()
+	if err != nil || catalogHref == "" {
+		return fmt.Errorf("empty catalog HREF for admin catalog %s", cat.AdminCatalog.Name)
+	}
+	err = cat.WaitForTasks()
+	if err != nil {
+		return err
+	}
+	return elementSync(cat.client, catalogHref, "admin catalog")
+}
+
+// LaunchSync starts synchronisation of a subscribed AdminCatalog
+func (cat *AdminCatalog) LaunchSync() (*Task, error) {
+	// if the catalog was not subscribed, return
+	if cat.AdminCatalog.ExternalCatalogSubscription == nil || cat.AdminCatalog.ExternalCatalogSubscription.Location == "" {
+		return nil, nil
+	}
+	// The sync operation is only available for Catalog, not AdminCatalog.
+	// We use the embedded Catalog object for this purpose
+	catalogHref, err := cat.GetCatalogHref()
+	if err != nil || catalogHref == "" {
+		return nil, fmt.Errorf("empty catalog HREF for admin catalog %s", cat.AdminCatalog.Name)
+	}
+	err = cat.WaitForTasks()
+	if err != nil {
+		return nil, err
+	}
+	return elementLaunchSync(cat.client, catalogHref, "admin catalog")
+}
+
+// GetCatalogHref retrieves the regular catalog HREF from an admin catalog
+func (cat *AdminCatalog) GetCatalogHref() (string, error) {
+	href := ""
+	for _, link := range cat.AdminCatalog.Link {
+		if link.Rel == "alternate" && link.Type == types.MimeCatalog {
+			href = link.HREF
+			break
+		}
+	}
+	if href == "" {
+		return "", fmt.Errorf("no regular Catalog HREF found for admin Catalog %s", cat.AdminCatalog.Name)
+	}
+	return href, nil
+}
+
+// QueryVappTemplateList returns a list of vApp templates for the given catalog
+func (catalog *AdminCatalog) QueryVappTemplateList() ([]*types.QueryResultVappTemplateType, error) {
+	return queryVappTemplateListWithFilter(catalog.client, map[string]string{"catalogName": catalog.AdminCatalog.Name})
+}
+
+// QueryMediaList retrieves a list of media items for the Admin Catalog
+func (catalog *AdminCatalog) QueryMediaList() ([]*types.MediaRecordType, error) {
+	return queryMediaList(catalog.client, catalog.AdminCatalog.HREF)
+}
+
+// LaunchSynchronisationVappTemplates starts synchronisation of a list of vApp templates
+func (cat *AdminCatalog) LaunchSynchronisationVappTemplates(nameList []string) ([]*Task, error) {
+	var taskList []*Task
+	for _, element := range nameList {
+		catalogItem, err := cat.QueryCatalogItem(element)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving catalog item %s: %s", element, err)
+		}
```

Comment:
A query by comparing behaviors of functions for vAppTemplates and MediaItems. It looks that functions that accept `nameList` are assymetric (not sure if this is a planned case)


I see that for `media` case we have such workflow:
`LaunchSynchronisationAllMediaItems` gathers list of names and passes it to `LaunchSynchronisationMediaItems`. `LaunchSynchronisationMediaItems` itself checks if each element has their tasks still running and waits for them.


For `vApp templates` it looks to be a bit different:

The function `LaunchSynchronisationAllVappTemplates` does checking for running tasks and waiting. It then calls `LaunchSynchronisationVappTemplates` with a name list. 

Can it be that if someone triggers `LaunchSynchronisationVappTemplates` without using `LaunchSynchronisationAllVappTemplates`, existing tasks are not verified like in the `LaunchSynchronisationMediaItems` case? Is there any purpose for this difference in behavior?

Reply:
It's a bit more complicated than that. The difference in the algorithms is because getting to the catalog item from the real entity is different from media and vApp template. The first function being implemented was the `LaunchSynchronisation[without-ALL}*`. Then I realised that some syncs were failing for existing tasks, and added the check. 
I was about to do the refactoring, when I was told that the fields that I was using to access the tasks in the `_all_` lists are not documented and could be dropped.
So, I deferred the refactoring, but I should redo all the `Sync` and `LaunchSync` functions without using the undocumented fields. It's a pity because any implementation that does not use the undocumented tasks needs several additional API calls to match items with tasks.
I am adding a comment and try to fix this problem before the release, either in this PR or in a new one.

Reply:
I changed the way these functions work, and they should now perform the same operations, albeit by different means